portal/src/util/validation.ts

```diff
@@ -0,0 +1,59 @@
+export interface ValidationRule<Data, ErrorState> {
+  inputKey: keyof Data;
+  errorKey: keyof ErrorState;
+  condition: (input: any) => boolean;
+  errorMessageId: string;
+}
+
+export type ErrorResult<ErrorState> = Partial<
+  { [key in keyof ErrorState]: string }
+>;
+
+export function validateInput<Data, ErrorState>(
+  data: Data,
+  validationRules: ValidationRule<Data, ErrorState>[]
+): {
+  errorResult: ErrorResult<ErrorState>;
+  isValid: boolean;
+} {
+  let isValid = true;
+  const errorResult: ErrorResult<ErrorState> = {};
+  validationRules.forEach(
+    (validationRule: ValidationRule<Data, ErrorState>) => {
+      if (!validationRule.condition(data[validationRule.inputKey])) {
+        errorResult[validationRule.errorKey] = validationRule.errorMessageId;
+      }
+      isValid = false;
+    }
+  );
+  return { errorResult, isValid };
+}
+
+export function isValidEmail(email: string): boolean {
```

Comment:
- An empty email should be invalid.
- Validating email through complex regex is frequently fragile. For example, newer TLD frequently has more than 3 characters and fails the regex. I would suggest just checking the presence of '@' character and non-empty local & domain part.